### PR TITLE
Add Http_lwt_client.fold_request for handling receiving streams of data instead of one large string that might not fit into RAM

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Feature request mentioned in https://github.com/roburio/http-lwt-client/issues/12#issuecomment-1248416483

### CODE REVIEW

**Comment by hannesm** [comment]
Thanks for your PR. I have two questions (didn't look into detail):
- should we at the same time have a way to specify the (potential) request body in the same shape (it shouldn't be required to fit into memory either)?
- the naming of the value. I've no good suggestion right now, but will think a bit more about it.

**Comment by kit-ty-kate** [comment]
> should we at the same time have a way to specify the (potential) request body in the same shape (it shouldn't be required to fit into memory either)?

Done. I haven't tested it and I'm no expert in the HTTP protocol header so please double-check my code

Feel free to change any part of the PR (as long as the overall feature, in whichever form, stays ofc ^^)

**Comment by kit-ty-kate** [comment]
Looking at `hurl` we might want to use `fold_request` here to avoid issues, or maybe simply remove (or rename `fold_request` to) `one_request` all together, as it makes it a bit too easy to shoot yourself in the foot (e.g. assume the response will fit into ram when it doesn’t).

But it depends if you’re fine breaking the API or not.

**Comment by hannesm** [comment]
Thanks for your additional changes (for the request body). Though from what I read, in HTTP/2 the "transfer-encoding: chunked" header is not allowed (see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) - but some other streaming mechanism should be applied (//cc @dinosaure who may have an idea?)

I'm fine breaking the API, but am slightly confused whether there should be "request body - Lwt_stream.t" and "response body - a custom fold API". Why not use a `Lwt_stream.t` in both cases?

**Comment by hannesm** [comment]
some notes, not yet ready (neither required for you to make these changes):
- the `one_request` function should take a `?body:string Lwt_stream.t -> ?body_length:int`, and the `content-length` header should be send with the value of `body_length` (if provided) -- AFAICT this header is not mandatory
- we're not using the transfer-encoding: chunked at all
- the variant whether None, a string, or a stream is sent, can be discarded (not afraid of API changes)
- the response body should as well be a string Lwt_stream.t that the client receives (and can then do what they like with it)

**Comment by kit-ty-kate** [comment]
> I'm fine breaking the API, but am slightly confused whether there should be "request body - Lwt_stream.t" and "response body - a custom fold API". Why not use a `Lwt_stream.t` in both cases?

I thought about it but the issues i was seeing is that:
- the expected behaviour of the function changes quite a bit. For example the function would now return even if the response hasn't finished being processed. If you're chaining calls back-to-back this can become a problem and a memory leak as well.
- I'm not sure how it would actually be done without exposing users to more memory leak. For example if you really don't care about the body, doing the simple `Http_lwt_client.request ... >= fun (response, _body) ->` you are still receiving some data in the background, just not using it and it creates more memory leak. Having a fold-like function forces the user to choose what to do with the data, instead of implicitly requiring the user to call `Lwt_stream.fold` afterwards anyway to avoid the memory leaks we're trying to avoid here.

But again that's just my intuition on how HTTP works

**Comment by kit-ty-kate** [comment]
> the `one_request` function should take a `?body:string Lwt_stream.t -> ?body_length:int`, and the `content-length` header should be send with the value of `body_length` (if provided) -- AFAICT this header is not mandatory

I feel like it's a bit weird of an API. For people who don't need to use streams but do need to provide a body (as i do myself in opam-health-check for example), it feels tedious to set up if we want to give a nice request to the server.

You go from a simple:
```
Http_lwt_client ~body:"the body" ...
```
to something like:
```
let body = "the body" in
let body_length = String length body in
let body = Lwt_stream.of_list [body] in
Http_lwt_client ~body ~body_length ...
```
it feels a bit clunky to use. Plus users will be forced to understand lwt_stream's behaviour, which can be a bit tricky.

Overall I think having ``?body:[`String of string | `Stream of string Lwt_stream.t]`` (or maybe with ` -> ?max_body_length:int` as well if you want) is a much easier API to use.

File: src/http_lwt_client.mli [hunk 1]
```diff
@@ -53,6 +57,28 @@ val pp_response : Format.formatter -> response -> unit
     The [happy-eyeballs] opam package is used to establish the TCP connection,
     which prefers IPv6 over IPv4.
 *)
+val fold_request
```
Comment (hannesm) [inline src/http_lwt_client.mli]:
```suggestion
val request
```

**Comment by hannesm** [comment]
TODO:
- drop `one_request` (to have a single function in the API)
- develop some tests (either by using cohttp, or as a binary download from the internet (not running with opam))

**Comment by hannesm** [comment]
I pushed a commit that removes the one_request, and adapts hurl with the changes.

The issue I ran into is that a POST with streaming/chunking does not work, neither with HTTP1 nor with HTTP2. The error for HTTP1 is `hurl: exception here: Failure("cannot write to closed writer")`; the error with HTTP2 is `hurl.exe: [ERROR] error PROTOCOL_ERROR (0x1):`.

I used https://ptsv2.com/t/wb83s-1666717614 as a test, and am unsure where to go now. Since I'm not eager to debug httpaf and h2, I'd move on and revert to "only the reply body is streamed", and once the request body being streamed is required, re-think how to achieve this. Any opinion (or insight)?

**Comment by hannesm** [comment]
I force-pushed the original change: the response body is a stream, and no longer a string. When CI passed, IMHO this is ready to go. As side, the input and output in hurl were switched (now fixed).

//cc @kit-ty-kate please approve or disapprove this change.

**Comment by kit-ty-kate** [comment]
Thanks for the merge. Looks good to me!

**Pull request merged**